libyul/backends/wasm/WasmCodeTransform.cpp

```diff
@@ -346,13 +361,29 @@ string WasmCodeTransform::newLabel()
 	return m_nameDispenser.newName("label_"_yulstring).str();
 }
 
-void WasmCodeTransform::allocateGlobals(size_t _amount)
+vector<size_t> WasmCodeTransform::allocateGlobals(vector<wasm::Type> const& _typesForGlobals)
 {
-	while (m_globalVariables.size() < _amount)
-		m_globalVariables.emplace_back(wasm::GlobalVariableDeclaration{
-			m_nameDispenser.newName("global_"_yulstring).str(),
-			wasm::Type::i64
-		});
+	vector<size_t> allocatedIndices;
+
+	map<wasm::Type, size_t> firstUnused;
```

Comment:
The name of this variable confuses me. We do not mark globals as used or unused, do we? It just matters if we have enough globals of the respective type, right?

Reply:
We don't mark them in a persistent way but the function now tells the caller which specific variables of each type can be used. This is necessary because `m_globalVariables` doesn't necessarily consist of two contiguous blocks of `i32` and `i64` variables. New globals are always appended at the end so can be a jumbled mix.

The loop is going over a list of types corresponding to output parameters and for each one it grabs the next unassigned global of that type. `firstUnused` just points at the next unassigned global (or, to be precise, the first global after the last assigned global of each type).

Maybe it would be clearer if I iterated over globals of each type separately? It would require two passes through `_typesForGlobals` but I could then replace `firstUnused` with a simple loop counter.

Reply:
I have just added a fixup that rewrites this function. The new code is longer, and slightly less efficient but now it does its job in discrete steps: count globals, count requested types, allocate new globals if necessary, assign global indices to variables.

If it's OK, let me know and I'll squash it with the original commit. If not, I can easily remove it.